cmd/fs-v1-multipart.go

```diff
@@ -18,324 +18,181 @@ package cmd
+)
 
-	return fsRemoveMeta(multipartBucketPath, uploadsMetaPath, tmpDir)
+// Returns EXPORT/.minio.sys/multipart/SHA256/UPLOADID
+func (fs *fsObjects) getUploadIDDir(bucket, object, uploadID string) string {
+	return pathJoin(fs.fsPath, minioMetaMultipartBucket, getSHA256Hash([]byte(pathJoin(bucket, object))), uploadID)
 }
 
-// Removes the uploadID, called either by CompleteMultipart of AbortMultipart. If the resuling uploads
-// slice is empty then we remove/purge the file.
-func (fs fsObjects) removeUploadID(bucket, object, uploadID string, rwlk *lock.LockedFile) (bool, error) {
-	uploadIDs := uploadsV1{}
-	_, err := uploadIDs.ReadFrom(rwlk)
-	if err != nil {
-		return false, err
-	}
-
-	// Removes upload id from the uploads list.
-	uploadIDs.RemoveUploadID(uploadID)
-
-	// Check this is the last entry.
-	if uploadIDs.IsEmpty() {
-		// No more uploads left, so we delete `uploads.json` file.
-		return true, fs.deleteUploadsJSON(bucket, object, uploadID)
-	} // else not empty
-
-	// Write update `uploads.json`.
-	_, err = uploadIDs.WriteTo(rwlk)
-	return false, err
+// Returns EXPORT/.minio.sys/multipart/SHA256
+func (fs *fsObjects) getMultipartSHADir(bucket, object string) string {
+	return pathJoin(fs.fsPath, minioMetaMultipartBucket, getSHA256Hash([]byte(pathJoin(bucket, object))))
 }
 
-// Adds a new uploadID if no previous `uploads.json` is
-// found we initialize a new one.
-func (fs fsObjects) addUploadID(bucket, object, uploadID string, initiated time.Time, rwlk *lock.LockedFile) error {
-	uploadIDs := uploadsV1{}
-
-	_, err := uploadIDs.ReadFrom(rwlk)
-	// For all unexpected errors, we return.
-	if err != nil && errors.Cause(err) != io.EOF {
-		return err
-	}
-
-	// If we couldn't read anything, we assume a default
-	// (empty) upload info.
-	if errors.Cause(err) == io.EOF {
-		uploadIDs = newUploadsV1("fs")
-	}
-
-	// Adds new upload id to the list.
-	uploadIDs.AddUploadID(uploadID, initiated)
-
-	// Write update `uploads.json`.
-	_, err = uploadIDs.WriteTo(rwlk)
-	return err
+// Returns partNumber.etag
+func (fs *fsObjects) encodePartFile(partNumber int, etag string) string {
+	return fmt.Sprintf("%.5d.%s", partNumber, etag)
 }
 
-// listMultipartUploadIDs - list all the upload ids from a marker up to 'count'.
-func (fs fsObjects) listMultipartUploadIDs(bucketName, objectName, uploadIDMarker string, count int) ([]MultipartInfo, bool, error) {
-	var uploads []MultipartInfo
-
-	// Hold the lock so that two parallel complete-multipart-uploads
-	// do not leave a stale uploads.json behind.
-	objectMPartPathLock := fs.nsMutex.NewNSLock(minioMetaMultipartBucket, pathJoin(bucketName, objectName))
-	if err := objectMPartPathLock.GetRLock(globalListingTimeout); err != nil {
-		return nil, false, errors.Trace(err)
+// Returns partNumber and etag
+func (fs *fsObjects) decodePartFile(name string) (partNumber int, md5hex string, err error) {
+	result := strings.Split(name, ".")
+	if len(result) != 2 {
+		return -1, "", errUnexpected
```

Comment:
A special -1 part number seems unnecessary as valid part numbers start from 1 anyway. In addition go conventions-wise, when error is not nil, the other return values must be ignored.